Plugin-able dependency injection for adapters

MorganMoon opened the pull request:
This allows people to plug in their IoC dependency injection frameworks to work with adapters. This is something I've had to create for my own projects, and I figured it can be useful for the framework as a whole. It's out of the way so people who don't want custom DI for the adapters wont have to change a thing, and I created a "DefaultWeldContainer" that gets used by default that just does what was being done normally. When creating your own implementation of IWeldContainerIoc, if you want it to be used by weld, just mark the class with the [WeldContainer] attribute, and if you want to supply a specific instance of this class you can mark a static method with the same attribute that returns a IWeldContainerIoc (or an implementing class of it). If you have multiple classes marked with the attribute, it will output a warning in the console & use the first one it finds :)

Example usage using with Zenject:
 - Create the weld container:
![image](https://user-images.githubusercontent.com/10552305/36926157-e25d96c2-1e3b-11e8-8cb6-88f9e40a1e21.png)
 - Zenject installer:
![image](https://user-images.githubusercontent.com/10552305/36926174-f235a2ec-1e3b-11e8-88f1-af61e006c43b.png)
 - Inject into adapter like normal: 
![image](https://user-images.githubusercontent.com/10552305/36926193-148f8fc4-1e3c-11e8-9102-a792e8c5c04b.png)


Hope this helps :)

Commit:
Created plugin-able dependency injection for adapters

MorganMoon committed:
Removed weird non-static AdapterResolver stuff

MorganMoon committed:
Removed underscore in variable name to keep code style consistent

MorganMoon committed:
Added warning if there are multiple classes marked with the WeldContainer attribute

MorganMoon committed:
AdapterResolver is now marked as a static class

RoryDungan committed:
Merge branch 'master' into feature/AdapterPluginableDI

MorganMoon commented:
Hey, just wondering what the status of this PR is, are there any requested changes? Or is it just not something that fits into the scope of weld?

RoryDungan commented on UnityWeld/Binding/Internal/AdapterResolver.cs:
```diff
@@ -0,0 +1,63 @@
+                    .Where(mi => mi.GetCustomAttributes(typeof(WeldContainerAttribute), false).Any() && typeof(IWeldContainerIoC).IsAssignableFrom(mi.ReturnType));
+                if (useableMethods.Any())
+                {
+                    var methodInfo = useableMethods.First();
```
Instead of calling `if (usableMethods.Any())` and then `usableMethods.First()`, you could get `usableMethods.FirstOrDefault()` and check if the result is null. This would potentially make the code a bit more efficient because as it is we're enumerating over the `Where` twice.

RoryDungan commented on UnityWeld/Binding/Internal/AdapterResolver.cs:
```diff
@@ -0,0 +1,63 @@
+            {
+                if(containerTypesCount > 1)
+                {
+                    UnityEngine.Debug.LogWarningFormat("You have multiple classes marked with the WeldContainer Attribute, only one can be used!");
```
I think I'd actually prefer if this threw an exception, since it's undefined behaviour and failing fast would force users to fix the issue.

RoryDungan commented on UnityWeld/Binding/Internal/AdapterResolver.cs:
```diff
@@ -0,0 +1,63 @@
+            if (container == null)
+            {
+                container = new DefaultWeldContainer();
+            }
```
This whole method might be a bit simpler if it just retruened the new IWeldContainerIoC rather than setting the `container` field. If it worked that way the method could be referentially transparent, and you could remove this last `if` statement and some of the other `else` statements because every instance of `container = ...` could be replaced with `return ...`

RoryDungan commented on UnityWeld/Ioc/DefaultWeldContainer.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace UnityWeld.Ioc
+{
+    public class DefaultWeldContainer : IWeldContainerIoC
```
Should add a `/// <summary> ... </summary>` comment

RoryDungan commented on UnityWeld/Ioc/IWeldContainerIoc.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace UnityWeld.Ioc
+{
+    public interface IWeldContainerIoC
```
Should add a `/// <summary> ...` comment explaining how the interface works so that users of the API can implement it themselves.

RoryDungan commented on UnityWeld/Ioc/WeldContainerAttribute.cs:
```diff
@@ -0,0 +1,9 @@
+namespace UnityWeld.Ioc
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, Inherited = false)]
+    public class WeldContainerAttribute : Attribute
```
Should add a `/// <summary> ...` comment explaining when to use the attribute.

RoryDungan commented on UnityWeld/Binding/Internal/AdapterResolver.cs:
```diff
@@ -0,0 +1,63 @@
+namespace UnityWeld.Binding.Internal
+{
+    /// <summary>
+    /// Helper class for creating adapters
```
Maybe also explain that it attempts to find the weld container via the attribute and use that but falls back to just using the DefaultWeldContainer if it can't.

RoryDungan commented on UnityWeld/Binding/Internal/AdapterResolver.cs:
```diff
@@ -0,0 +1,63 @@
+
+        private static void SetupWeldContainer()
+        {
+            var containerTypes = TypeResolver.TypesWithWeldContainerAttribute.Where(t => typeof(IWeldContainerIoC).IsAssignableFrom(t));
```
Should call `.ToList()` on this to avoid multiple enumeration. If it is a list, the later call to Linq `.Count()` can be replaced with just the `.Count` property, which is O(1) instead of O(n) complexity.

RoryDungan reviewed (changes requested):
Sorry it's taken me so long to get to this, we've just been quite busy with other projects. The functionality is very useful and fixes an issue that has made working with more advanced kinds of adapters difficult in the past. There are just a few changes I think can be made to make the code quality a bit better.

Commit:
Making requested changes

- Reworked use of linq to be more efficient
- If there is more than one container marked by attribute, AdapterResolver throws an exception
- Added summary comments

MorganMoon commented:
No problem :) Made the requested changes

Commit:
T has to be a class in IWeldContainerIoC
